lib.strings: Prevent paths as inputs in some functions

infinisil opened the pull request:
###### Description of changes
Some functions in `lib.strings` could lead to unexpected results when path data types are passed. This is confusing with `lib.path` now existing and [being extended](https://github.com/NixOS/nixpkgs/issues/210426).

```
# Always returns false
nix-repl> lib.strings.hasPrefix ./lib ./lib
false

# No point, since absolute paths are always normalised, normalizePath is explicitly only useful for strings
nix-repl> lib.strings.normalizePath ./lib
"/nix/store/6d499zmgypxx73c7gdm3b36mkxy21iis-lib"
```

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

roberth commented on lib/strings.nix:
```diff
@@ -205,7 +205,12 @@ rec {
+    if isPath s then throw ''
+      lib.strings.normalizePath: The argument (${toString s}) is not a string but a path, which is not supported.
+          Values of the path data type are always normalised in Nix, there's no need to do that manually.
+          To normalise subpaths when working with path data types, please use `lib.path.subpath.normalise`.''
```
What's a _path data type_?

infinisil replied:
A path in the data type listing for Nix: https://nixos.org/manual/nix/stable/language/values.html?type-path#type-path

I find just "Paths are always normalised in Nix" a bit too confusing, given that we can have paths in strings too.

infinisil replied:
Now changed this to "Path values are always normalised in Nix", which is what you also used in your other suggestions, I think that sounds good.

roberth replied:
Sounds good. I had an inkling but my mind did a 180 at "data type" which is a bit outside the usual lingo.

roberth commented on lib/strings.nix:
```diff
@@ -205,7 +205,12 @@ rec {
+      lib.strings.normalizePath: The argument (${toString s}) is not a string but a path, which is not supported.
+          Values of the path data type are always normalised in Nix, there's no need to do that manually.
+          To normalise subpaths when working with path data types, please use `lib.path.subpath.normalise`.''
+    else builtins.foldl' (x: y: if y == "/" && hasSuffix "/" x then x else x+y) "" (stringToCharacters s);
```
```suggestion
    else
      builtins.foldl'
        (x: y: if y == "/" && hasSuffix "/" x then x else x+y)
        ""
        (stringToCharacters s);
```

Now we can read it, yay!

roberth commented on lib/strings.nix:
```diff
@@ -238,7 +243,12 @@ rec {
+    # str must be a string-like
+    if isPath pref then throw ''
+      lib.strings.hasPrefix: The first argument (${toString pref}) is a path, which is not supported''
+    else (substring 0 (stringLength pref) str == pref);
```
```suggestion
    else
      substring 0 (stringLength pref) str == pref;
```

Nothing fits on screen anyway because of the large doc strings. Might as well improve readability.

roberth commented on lib/strings.nix:
```diff
@@ -238,7 +243,12 @@ rec {
+    str:
+    # pref must be a string
+    # str must be a string-like
+    if isPath pref then throw ''
+      lib.strings.hasPrefix: The first argument (${toString pref}) is a path, which is not supported''
```
```suggestion
    # Before 23.05, strings would be added to the store before comparing. This was too unexpected.
    if isPath pref then throw ''
      lib.strings.hasPrefix: The first argument (${toString pref}) is a path value, but only strings are supported.''
```

Leave a clue in case users encounter this in their code base, so they know something has changed here. It appears in the code block of the stack trace iirc. The exception message does not include the whole history, because I don't expect that to be productive. It could stick around for years being unnecessarily noisy.

infinisil replied:
Sounds good, but "only strings are supported" isn't entirely correct, all `stringLike` values (including `outPath`, `__toString`) are supported, except paths. Maybe that's good enough though

roberth replied:
Right. Probably best to keep it simple and avoid unnecessary dissonance?
This can only be cured with poetry. Cue ChatGPT:

Here is a revised limerick about type coercions in programming:

Type coercion can be quite slick,
To convert data types that don't mix.
But if you're not careful,
Your code can be awful,
And bugs will give your code a big kick.

infinisil replied:
Haha nicely done :laughing:

I agree, let's keep it simple, I applied this suggestion to all the error messages (also the normalizePath one).

roberth commented on lib/strings.nix:
```diff
@@ -258,7 +268,10 @@ rec {
-    in lenContent >= lenSuffix &&
+    in
+    if isPath suffix then throw ''
+      lib.strings.hasSuffix: The first argument (${toString suffix}) is a path, which is not supported''
```
See `hasPrefix` comments.

roberth commented on lib/strings.nix:
```diff
@@ -276,7 +289,9 @@ rec {
   hasInfix = infix: content:
-    builtins.match ".*${escapeRegex infix}.*" "${content}" != null;
+    if isPath infix then throw ''
+      lib.strings.hasInfix: The first argument (${toString infix}) is a path, which is not supported''
```
Same.

infinisil committed:
lib.strings: Deprecate path prefix/suffix/infix arguments

lib.{hasPrefix,hasInfix,hasSuffix} would otherwise return an
always-false result, which can be very unexpected:

    nix-repl> lib.strings.hasPrefix ./lib ./lib/meta.nix
    false

infinisil committed:
lib.strings.normalizePath: Deprecate for path values

There's no need to call this function on path data types, and it's
confusing with the new lib.path library functions

fricklerhandwerk commented on lib/strings.nix:
```diff
@@ -238,7 +247,12 @@ rec {
     # Input string
-    str: substring 0 (stringLength pref) str == pref;
+    str:
+    # Before 23.05, strings would be added to the store before comparing. This was too unexpected.
```
```suggestion
    # Before 23.05, paths would be copied to the store before converting them to strings and comparing. This was surprising and confusing.
```

Also the other instance

fricklerhandwerk reviewed (commented):
This is a breaking change and at the very least needs release notes. Almost certainly this will error out on someone in production. We may even want to phase this in with a warning or setting. Who else should we ask to make that decision? @roberth opinions?

infinisil commented:
Hmm yeah I guess it's a breaking change, even if the code we break is certainly a bug (in case of `hasPrefix`), or unnecessary (in case of `normalizePath`).

I changed the errors to a warning instead and added some more explanatory text to them.

fricklerhandwerk commented on lib/strings.nix:
```diff
@@ -205,7 +207,20 @@ rec {
+      ''
+        lib.strings.normalizePath: The argument (${toString s}) is a path value, but only strings are supported.
+            Path values are always normalised in Nix, so there's no need to call this function on them.
+            This function also imports the path into the Nix store and returns the store path, the same as "''${path}" will, which may not be desired.
```
```suggestion
            This function also copied the path to the Nix store and returns the store path, the same as "''${path}" will, which may not be what you want.
```

Import is a loaded term around here, and may produce the wrong associations. Let's stick to more everyday words.

Desired - by whom? Passive voice sometimes obscures what you may want to say. What is the desired message here?

infinisil replied:
Applied :+1:

fricklerhandwerk commented:
Have to take the space bar heaters into account... been there myself.

infinisil committed:
lib.strings.remove{Prefix,Suffix}: Deprecate for path prefix/suffix arguments

See also parent commits

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-47/27387/1
